File: transports/elixir/lib/csilgen/transport/events.ex
defmodule Csilgen.Transport.Events do
  @moduledoc """
  CSIL-Events transport — typed bidirectional event streams — see
  csil-events-transport.md. Verbose (text-keyed) and compact (positional)
  profiles, plus the control-plane (service ordinal 0) lifecycle payloads.
  """

  alias Csilgen.Transport.{CBOR, Conventions}

  defmodule Event do
    @moduledoc """
    One typed event flowing in either direction. Identified by service+operation
    (verbose) or their ordinals (compact); carries an optional correlation id.
    """
    defstruct [:service, :service_ord, :event, :op_ord, :id, payload: <<>>]

    @type t :: %__MODULE__{
            service: String.t() | nil,
            service_ord: non_neg_integer() | nil,
            event: String.t() | nil,
            op_ord: non_neg_integer() | nil,
            id: non_neg_integer() | nil,
            payload: binary()
          }
  end

  @doc "Builds a verbose event by name."
  @spec new_verbose(String.t() | nil, String.t(), binary()) :: Event.t()
  def new_verbose(service, event, payload) do
    %Event{service: service, event: event, payload: payload}
  end

  @doc "Builds a compact event by ordinals."
  @spec new_compact(non_neg_integer(), non_neg_integer(), binary()) :: Event.t()
  def new_compact(service_ord, op_ord, payload) do
    %Event{service_ord: service_ord, op_ord: op_ord, payload: payload}
  end

  @doc "Parses a wire profile name."
  @spec parse_profile(String.t()) :: {:ok, :verbose | :compact} | :error
  def parse_profile("verbose"), do: {:ok, :verbose}
  def parse_profile("compact"), do: {:ok, :compact}
  def parse_profile(_), do: :error

  @spec encode_event(Event.t(), :verbose | :compact) :: binary()
  def encode_event(%Event{} = e, :verbose) do
    entries =
      [{"event", e.event}, {"payload", Conventions.tag24(e.payload)}]
      |> maybe_put("service", e.service)
      |> maybe_put("id", e.id)

    CBOR.encode({:map, entries})
  end

  def encode_event(%Event{} = e, :compact) do
    arr = [e.service_ord, e.op_ord]
    arr = if e.id != nil, do: arr ++ [e.id], else: arr
    CBOR.encode(arr ++ [Conventions.tag24(e.payload)])
  end

  @spec decode_event(binary(), :verbose | :compact) :: {:ok, Event.t()} | {:error, term()}
  def decode_event(bytes, :verbose) do
    with {:ok, map} <- Conventions.decode_envelope(bytes),
         {:ok, payload} <- Conventions.get_payload(map),
         {:ok, event} <- Conventions.get_text(map, "event") do
      {:ok,
       %Event{
         service: Conventions.get_text_opt(map, "service"),
         event: event,
         id: Conventions.get_uint_opt(map, "id"),
         payload: payload
       }}
    end
  end

  def decode_event(bytes, :compact) do
    with {:ok, arr} when is_list(arr) <- Conventions.decode_envelope(bytes) do
      # 3 elements => [service_ord, op_ord, payload]; 4 => with correlation id.
      case arr do
        [so, oo, payload_v] ->
          build_compact(so, oo, nil, payload_v)

        [so, oo, id_v, payload_v] ->
          build_compact(so, oo, id_v, payload_v)

        _ ->
          {:error, :bad_compact_arity}
      end
    end
  end

  defp build_compact(so, oo, id_v, payload_v)
       when is_integer(so) and is_integer(oo) do
    with {:ok, payload} <- Conventions.untag24(payload_v) do
      {:ok, %Event{service_ord: so, op_ord: oo, id: id_v, payload: payload}}
    end
  end

  defp build_compact(_so, _oo, _id, _payload), do: {:error, :non_integer_ordinal}

  # --- Control plane --------------------------------------------------------

  defmodule Hello do
    @moduledoc "The `$hello` payload offered by the connection initiator."
    @enforce_keys [:versions, :profiles]
    defstruct [:versions, :profiles, :service, :auth]

    @type t :: %__MODULE__{
            versions: [non_neg_integer()],
            profiles: [String.t()],
            service: String.t() | nil,
            auth: String.t() | nil
          }
  end

  defmodule HelloAck do
    @moduledoc "The `$hello-ack` payload returned by the peer."
    @enforce_keys [:v, :profile]
    defstruct [:v, :profile, :session]
    @type t :: %__MODULE__{v: non_neg_integer(), profile: String.t(), session: String.t() | nil}
  end

  defmodule Heartbeat do
    @moduledoc "A `$ping`/`$pong` heartbeat payload."
    @enforce_keys [:nonce]
    defstruct [:nonce, :at]
    @type t :: %__MODULE__{nonce: non_neg_integer(), at: non_neg_integer() | nil}
  end

  defmodule Close do
    @moduledoc "A `$close` payload."
    @enforce_keys [:status]
    defstruct [:status, :reason]
    @type t :: %__MODULE__{status: integer(), reason: String.t() | nil}
  end

  @spec encode_hello(Hello.t()) :: binary()
  def encode_hello(%Hello{} = h) do
    entries =
      [{"versions", h.versions}, {"profiles", h.profiles}]
      |> maybe_put("service", h.service)
      |> maybe_put("auth", h.auth)

    CBOR.encode({:map, entries})
  end

  @spec encode_hello_ack(HelloAck.t()) :: binary()
  def encode_hello_ack(%HelloAck{} = a) do
    entries =
      [{"v", a.v}, {"profile", a.profile}]
      |> maybe_put("session", a.session)

    CBOR.encode({:map, entries})
  end

  @spec encode_heartbeat(Heartbeat.t()) :: binary()
  def encode_heartbeat(%Heartbeat{} = h) do
    entries = [{"nonce", h.nonce}] |> maybe_put("at", h.at)
    CBOR.encode({:map, entries})
  end

  @spec encode_close(Close.t()) :: binary()
  def encode_close(%Close{} = c) do
    entries = [{"status", c.status}] |> maybe_put("reason", c.reason)
    CBOR.encode({:map, entries})
  end

  defp maybe_put(entries, _key, nil), do: entries
  defp maybe_put(entries, key, value), do: entries ++ [{key, value}]
end

File: transports/ruby/lib/csilgen/transport/events.rb
# frozen_string_literal: true

module Csilgen
  module Transport
    # CSIL-Events transport -- typed bidirectional event streams -- see
    # csil-events-transport.md. Verbose (text-keyed) and compact (positional) profiles,
    # plus the control plane (service ordinal 0) lifecycle payloads.
    module Events
      # Which wire profile a connection uses for its lifetime, fixed by `hello`.
      # Symbols are the lightest idiomatic enum in Ruby.
      module Profile
        VERBOSE = :verbose
        COMPACT = :compact
        ALL = [VERBOSE, COMPACT].freeze

        module_function

        def parse(text)
          case text
          when "verbose" then VERBOSE
          when "compact" then COMPACT
          end
        end

        def as_str(profile)
          profile.to_s
        end
      end

      # One typed event flowing in either direction. Identified by service+operation
      # (verbose) or by their ordinals (compact); carries an optional correlation `id`
      # when it is a request expecting a reply, or that reply. The compact ordinals are
      # full CBOR uints (events doc section 2), unlike the fixed-width datagram header.
      Event = ::Data.define(:service, :service_ord, :event, :op_ord, :id, :payload) do
        def initialize(service: nil, service_ord: nil, event: nil, op_ord: nil, id: nil, payload: "".b)
          super
        end

        def self.verbose(service, event, payload)
          new(service: service, event: event, payload: payload)
        end

        def self.compact(service_ord, op_ord, payload)
          new(service_ord: service_ord, op_ord: op_ord, payload: payload)
        end

        def with_id(new_id) = with(id: new_id)

        def encode(profile)
          (profile == Profile::VERBOSE) ? encode_verbose : encode_compact
        end

        def encode_verbose
          raise MalformedError, "verbose event missing 'event' name" if event.nil?

          entries = [
            ["event", event],
            ["payload", Conventions.tag24(payload)]
          ]
          entries << ["service", service] unless service.nil?
          entries << ["id", id] unless id.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def encode_compact
          raise MalformedError, "compact event missing service ordinal" if service_ord.nil?
          raise MalformedError, "compact event missing op ordinal" if op_ord.nil?

          arr = [service_ord, op_ord]
          arr << id unless id.nil?
          arr << Conventions.tag24(payload)
          Conventions.encode_value(arr)
        end

        def self.decode(data, profile)
          (profile == Profile::VERBOSE) ? decode_verbose(data) : decode_compact(data)
        end

        def self.decode_verbose(data)
          value = Conventions.decode_value(data)
          payload_value = Conventions.map_get(value, "payload")
          raise MalformedError, "missing 'payload'" if payload_value.nil?

          new(
            service: Conventions.get_text_opt(value, "service"),
            event: Conventions.get_text(value, "event"),
            id: Conventions.get_uint_opt(value, "id"),
            payload: Conventions.untag24(payload_value)
          )
        end

        def self.decode_compact(data)
          value = Conventions.decode_value(data)
          raise MalformedError, "compact event is not an array" unless value.is_a?(::Array)

          # 3 elements => [service_ord, op_ord, payload]; 4 => with correlation id.
          case value.length
          when 3
            service_ord, op_ord, payload_value = value
            id_value = nil
          when 4
            service_ord, op_ord, id_value, payload_value = value
          else
            raise MalformedError, "compact event array has #{value.length} elements, expected 3 or 4"
          end

          as_uint = lambda do |val|
            raise MalformedError, "ordinal is not a non-negative integer" unless val.is_a?(::Integer) && val >= 0

            val
          end

          new(
            service_ord: as_uint.call(service_ord),
            op_ord: as_uint.call(op_ord),
            id: id_value.nil? ? nil : as_uint.call(id_value),
            payload: Conventions.untag24(payload_value)
          )
        end
      end

      # Control-plane operation ordinals (under service ordinal 0) and their `$`-sigil
      # verbose names.
      module Control
        HELLO = 0
        HELLO_ACK = 1
        PING = 2
        PONG = 3
        CLOSE = 4
        ERROR = 5

        HELLO_NAME = "$hello"
        HELLO_ACK_NAME = "$hello-ack"
        PING_NAME = "$ping"
        PONG_NAME = "$pong"
        CLOSE_NAME = "$close"
        ERROR_NAME = "$error"
      end

      # The `$hello` payload offered by the connection initiator.
      Hello = ::Data.define(:versions, :profiles, :service, :auth) do
        def initialize(versions:, profiles:, service: nil, auth: nil)
          super
        end

        def encode
          entries = [
            ["versions", versions.dup],
            ["profiles", profiles.dup]
          ]
          entries << ["service", service] unless service.nil?
          entries << ["auth", auth] unless auth.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          versions_value = Conventions.map_get(value, "versions")
          raise MalformedError, "hello missing 'versions'" unless versions_value.is_a?(::Array)

          versions = versions_value.select { |v| v.is_a?(::Integer) && v >= 0 }
          profiles_value = Conventions.map_get(value, "profiles")
          raise MalformedError, "hello missing 'profiles'" unless profiles_value.is_a?(::Array)

          profiles = profiles_value.select { |p| p.is_a?(::String) }
          new(
            versions: versions,
            profiles: profiles,
            service: Conventions.get_text_opt(value, "service"),
            auth: Conventions.get_text_opt(value, "auth")
          )
        end

        # Select a profile from this hello's offers, honoring the peer's preference
        # order and what the server supports. Returns the chosen [version, profile], or
        # nil if nothing is mutually supported.
        def negotiate(supported)
          return nil unless versions.include?(Conventions::VERSION)

          profiles.each do |offered|
            parsed = Profile.parse(offered)
            return [Conventions::VERSION, parsed] if parsed && supported.include?(parsed)
          end
          nil
        end
      end

      # The `$hello-ack` payload returned by the peer.
      HelloAck = ::Data.define(:v, :profile, :session) do
        def initialize(v:, profile:, session: nil)
          super
        end

        def encode
          entries = [
            ["v", v],
            ["profile", profile]
          ]
          entries << ["session", session] unless session.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          new(
            v: Conventions.get_uint(value, "v"),
            profile: Conventions.get_text(value, "profile"),
            session: Conventions.get_text_opt(value, "session")
          )
        end
      end

      # A `$ping` / `$pong` heartbeat payload.
      Heartbeat = ::Data.define(:nonce, :at) do
        def initialize(nonce:, at: nil)
          super
        end

        def encode
          entries = [["nonce", nonce]]
          entries << ["at", at] unless at.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          new(nonce: Conventions.get_uint(value, "nonce"), at: Conventions.get_uint_opt(value, "at"))
        end
      end

      # A `$close` payload.
      Close = ::Data.define(:status, :reason) do
        def initialize(status:, reason: nil)
          super
        end

        def encode
          entries = [["status", status.code]]
          entries << ["reason", reason] unless reason.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          new(
            status: Status.from_code(Conventions.get_int(value, "status")),
            reason: Conventions.get_text_opt(value, "reason")
          )
        end
      end
    end
  end
end
